src/v2/helpers.rs

```diff
@@ -168,6 +168,82 @@ where
+    deserializer: D,
+) -> Result<BTreeMap<String, Option<RawOr<String>>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    /// Declare an internal visitor type to handle our input.
+    struct MapOrKeyOptionalValueListVisitor;
+
+    impl<'de> Visitor<'de> for MapOrKeyOptionalValueListVisitor {
+        type Value = BTreeMap<String, Option<RawOr<String>>>;
+
+        // We have a real map.
+        fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+        where
+            V: MapAccess<'de>,
+        {
+            let mut map: BTreeMap<String, Option<RawOr<String>>> = BTreeMap::new();
+            while let Some(key) = visitor.next_key::<String>()? {
+                if map.contains_key(&key) {
+                    let msg = format!("duplicate map key: {}", &key);
+                    return Err(<V::Error as de::Error>::custom(msg));
+                }
+                let ConvertToString(val) = visitor.next_value::<ConvertToString>()?;
+                let raw_or_value = raw(val)
+                    .map_err(|e| <V::Error as de::Error>::custom(format!("{}", e)))?;
+                map.insert(key, Some(raw_or_value));
+            }
+            Ok(map)
+        }
+
+        // We have a key/value list.  Yuck.
+        fn visit_seq<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+        where
+            V: SeqAccess<'de>,
+        {
+            lazy_static! {
+                // Match a key/value pair.
+                static ref KEY_VALUE: Regex =
+                    Regex::new("^([^=]+)(=(.*))?$").unwrap();
+            }
+
+            let mut map: BTreeMap<String, Option<RawOr<String>>> = BTreeMap::new();
+            while let Some(key_value) = visitor.next_element::<String>()? {
+                let caps = KEY_VALUE.captures(&key_value).ok_or_else(|| {
+                    let msg = format!("expected KEY[=value], got: <{}>", &key_value);
+                    <V::Error as de::Error>::custom(msg)
+                })?;
+                let key = caps.get(1).unwrap().as_str();
+                let value = caps.get(3).map(|v| v.as_str());
+                if map.contains_key(key) {
+                    let msg = format!("duplicate map key: {}", key);
+                    return Err(<V::Error as de::Error>::custom(msg));
+                }
+                let optional_raw_or_value = match value.map(|value| {
+                    raw(value.to_owned())
+                        .map_err(|e| <V::Error as de::Error>::custom(format!("{}", e)))
+                }) {
+                    None => None,
+                    Some(x) => Some(x?),
+                };
+                map.insert(key.to_owned(), optional_raw_or_value);
+            }
+            Ok(map)
+        }
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            write!(formatter, "a map or a key/[value] list")
+        }
+    }
+
+    deserializer.deserialize_map(MapOrKeyOptionalValueListVisitor)
```

Comment:
Just last night, I discovered in https://github.com/serde-rs/serde/issues/1125#issuecomment-353151077 that there has been a breaking change to the `serde` crate, and it is no longer safe to call `deserialize_map` as a hint when when override both `visit_map` and `visit_seq`. This needs to be `deserialize_any` (and I need to fix this everywhere else in `compose_yml` as well). Sorry for this last minute surprise. :-(

Reply:
I fixed this, and in the other obvious places too.